kernel/msg_q.c

```diff
@@ -166,23 +178,29 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
 		/* don't wait for message space to become available */
 		result = -ENOMSG;
 	} else {
-		SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_msgq, put, msgq, timeout);
+		SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_msgq, put_internal, msgq, timeout);
 
 		/* wait for put message success, failure, or timeout */
 		_current->base.swap_data = (void *) data;
 
 		result = z_pend_curr(&msgq->lock, key, &msgq->wait_q, timeout);
-		SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, result);
+		SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put_internal, msgq, timeout, result);
 		return result;
 	}
 
-	SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, result);
+	SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put_internal, msgq, timeout, result);
 
 	k_spin_unlock(&msgq->lock, key);
 
+out:
 	return result;
 }
 
+int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
+{
+	return z_impl_k_msgq_put_internal(msgq, data, timeout, false);
```

Comment:
I would guess that the sys_trace macros should probably go in the wrapper functions, because those are the externally visible API calls that people will expect to trace.

Reply:
Do you mean add a api called `sys_port_trace_k_msgq_put_front_blocking` to `z_impl_k_msgq_put_front`?
If so, there is five lines need to be modified for `SYS_PORT_TRACING_OBJ_FUNC_XXX` and these lines is distributed in `z_impl_k_msgq_put`.

There is a question which is I don't have good ideal to deal the same logic in `z_impl_k_msgq_put` and `z_impl_k_msgq_put_front`.

Reply:
(I take no position on tracing.  Honestly I think our current tracing layer is sort of a kludge and constantly breaking as code gets moved around.  Tracing done well needs to happen at a lower level, like automatic toolchain annotation or dtrace-style dynamic hooking.)

Reply:
> Do you mean add a api called `sys_port_trace_k_msgq_put_front_blocking` to `z_impl_k_msgq_put_front`?
> If so, there are five lines need to be modified for `SYS_PORT_TRACING_OBJ_FUNC_XXX` and these lines is distributed in `z_impl_k_msgq_put`.

The call to the tracing macros need to reference the the action being called (put or prepend), so in case of put you will be calling `SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_msgq, put, msgq, timeout);` for example, in the case of prepending, you will call `SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_msgq, prepend, msgq, timeout);`

Reply:
> tracing

I briefly looked at how another rtos does dynamic probes but it it looks like at least in that instance it'd require setting hardware break points and triggering NMIs, and deciphering some state of where the program was before the interrupt, recording it, and returning. Not saying its not clever or useful but compared to the macros we have today... pretty heavy handed in terms of over head.

I *do* love the idea that we could arbitrarily add probes like that though, its clever and cool for sure.